File: WinTexRFID/Backup/custom/m775Authenticate/M775AuthenticateForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using BLEDeviceAPI;

namespace UHFAPP.custom.m775Authenticate
{
    public partial class M775AuthenticateForm : Form
    {
        public M775AuthenticateForm()
        {
            InitializeComponent();
        }
        M775AuthenticateAPI api = new M775AuthenticateAPI();

        private void btnAuthenticate_Click(object sender, EventArgs e)
        {
            if (!DetectionFiltration())
            {
                return;
            }
             txtAuthenticateData.Text = "00000000000000000000";

            if (txtAuthenticateData.Text.Replace(" ", "").Length != 20)
            {
                MessageBox.Show("The length of the message must be 10 bytes!");
                return;
            }

            byte filterBank = 1;
            if (rbTID.Checked)
            {
                filterBank = 2;
            }
            else if (rbUser.Checked)
            {
                filterBank = 3;
            }
            int filterAddr = int.Parse(txtPtr.Text);
            int filterDataLen = int.Parse(txtLen.Text);
            byte[] filterData = DataConvert.HexStringToByteArray(txtFilter_EPC.Text);
            byte keyID = 0;
            byte[] tData = DataConvert.HexStringToByteArray(txtAuthenticateData.Text);
            int password = 0;

            byte[] result = api.UHFAuthenticate(password, filterBank, filterAddr, filterDataLen, filterData, keyID, tData);
            if (result == null || result.Length == 0)
            {
                MessageBox.Show("fail");
                txtTid.Text = "";
                return;
            }
            else if (result.Length!=22)
            {
                MessageBox.Show("返回数据长度错误");
                txtTid.Text = "";
                return;
            }
            //Challenge：6个字节，Tags Shortened TID：8个字节，Tag Response：8个字节

            txtChallenge.Text = DataConvert.ByteArrayToHexString(Utils.CopyArray(result, 0, 6));
            txtTid.Text = DataConvert.ByteArrayToHexString(Utils.CopyArray(result, 6, 8));
            txtResponse.Text = DataConvert.ByteArrayToHexString(Utils.CopyArray(result, 14, 8));
           // MessageBox.Show("success");
        }

        private void txtFilter_EPC_TextChanged(object sender, EventArgs e)
        {
            FormatHex(txtFilter_EPC);
            string data = txtFilter_EPC.Text.Replace(" ", "");
            if (data.Length > 0)
            {
                txtFilter_EPCLen.Text = ((data.Length / 2) + ((data.Length % 2) == 0 ? 0 : 1)).ToString();  // txtRead_Length.Text = ((data.Length / 4) + ((data.Length % 4) == 0 ? 0 : 1)).ToString();
            }
            else
            {
                txtFilter_EPCLen.Text = "0";
            } 
        }
        private void FormatHex(TextBox txt)
        {
            if (isDelete) return;
            string data = txt.Text.Trim().Replace(" ", "");
            if (data != string.Empty)
            {
                int SelectIndex = txt.SelectionStart - 1;
                char[] charData = data.ToCharArray(0, data.Length);
                char[] newChar = new char[charData.Length];
                int index = 0;
                for (int k = 0; k < charData.Length; k++)
                {
                    if (StringUtils.IsHexNumber2(charData[k]))
                    {
                        newChar[index] = charData[k];
                        index++;
                    }
                }
                string newData = new string(newChar, 0, index);
                StringBuilder sb = new StringBuilder();
                int count = (newData.Length / 2) + (newData.Length % 2);

                for (int k = 0; k < count; k++)
                {
                    if ((k * 2 + 2) <= newData.Length)
                    {
                        sb.Append(newData.Substring(k * 2, 2));
                    }
                    else
                    {
                        sb.Append(newData.Substring(k * 2, 1));
                    }
                    sb.Append(" ");
                }
                txt.Text = sb.ToString();

                if (SelectIndex >= 0)
                    txt.SelectionStart = SelectIndex + 2;
                //txt.Select(txt.Text.Length - 1, 0);

            }
        }
        bool isDelete = false;

        private void M775AuthenticateForm_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Back)
            {
                isDelete = true;
            }
            else
            {
                isDelete = false;
            }
        }
        private string HexFormat(string hex)
        {
            try
            {
                StringBuilder sb = new StringBuilder();
                for (int k = 0; k < hex.Length / 2; k++)
                {
                    sb.Append(hex.Substring(k * 2, 2));
                    if (k < hex.Length / 2 - 1)
                    {
                        sb.Append(" ");
                    }
                }
                return sb.ToString().ToUpper();
            }
            catch (System.Exception ex)
            {
                return string.Empty;
            }

        }

        private bool DetectionFiltration()
        {

            if (txtLen.Text.Trim().Length != 0)
            {
                if (int.Parse(txtLen.Text) > 0)
                {
                    string filerData = txtFilter_EPC.Text.Replace(" ", "");
                    if (!StringUtils.IsHexNumber(filerData))
                    {
                        MessageBox.Show(Common.isEnglish ? "Please input the hex filter data!" : "请输入十六进制过滤数据!");
                        return false;
                    }

                    byte[] filerBuff = DataConvert.HexStringToByteArray(filerData);
                    int filerPtr = int.Parse(txtPtr.Text);
                    int filerLen = int.Parse(txtLen.Text);

                    if ((filerLen / 8 + (filerLen % 8 == 0 ? 0 : 1)) * 2 > filerData.Length)
                    {
                        MessageBox.Show(Common.isEnglish ? "filter data length error!" : "过滤数据和长度不匹配!");  //to do
                        return false;
                    }
                }

            }
            return true;
        }
        private void rbEPC_Click(object sender, EventArgs e)
        {
            txtPtr.Text = "32";
        }

        private void rbTID_Click(object sender, EventArgs e)
        {
            txtPtr.Text = "0";
        }

        private void rbUser_Click(object sender, EventArgs e)
        {
            txtPtr.Text = "0";
        }

        private void txtAuthenticateData_TextChanged(object sender, EventArgs e)
        {
            FormatHex(txtAuthenticateData);
            string data = txtAuthenticateData.Text.Replace(" ", "");
            if (data.Length > 0)
            {
                txtDataLen.Text = ((data.Length / 2) + ((data.Length % 2) == 0 ? 0 : 1)).ToString();  // txtRead_Length.Text = ((data.Length / 4) + ((data.Length % 4) == 0 ? 0 : 1)).ToString();
            }
            else
            {
                txtDataLen.Text = "0";
            }
        }
    }
}

File: WinTexRFID/custom/m775Authenticate/M775AuthenticateForm.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports BLEDeviceAPI

Namespace UHFAPP.custom.m775Authenticate
	Partial Public Class M775AuthenticateForm
		Inherits Form

		Public Sub New()
			InitializeComponent()
		End Sub
		Private api As New M775AuthenticateAPI()

		Private Sub btnAuthenticate_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnAuthenticate.Click
			If Not Me.btnAuthenticate.IsHandleCreated Then Return

			If Not DetectionFiltration() Then
				Return
			End If
			 txtAuthenticateData.Text = "00000000000000000000"

			If txtAuthenticateData.Text.Replace(" ", "").Length <> 20 Then
				MessageBox.Show("The length of the message must be 10 bytes!")
				Return
			End If

			Dim filterBank As Byte = 1
			If rbTID.Checked Then
				filterBank = 2
			ElseIf rbUser.Checked Then
				filterBank = 3
			End If
			Dim filterAddr As Integer = Integer.Parse(txtPtr.Text)
			Dim filterDataLen As Integer = Integer.Parse(txtLen.Text)
			Dim filterData() As Byte = DataConvert.HexStringToByteArray(txtFilter_EPC.Text)
			Dim keyID As Byte = 0
			Dim tData() As Byte = DataConvert.HexStringToByteArray(txtAuthenticateData.Text)
			Dim password As Integer = 0

			Dim result() As Byte = api.UHFAuthenticate(password, filterBank, filterAddr, filterDataLen, filterData, keyID, tData)
			If result Is Nothing OrElse result.Length = 0 Then
				MessageBox.Show("fail")
				txtTid.Text = ""
				Return
			ElseIf result.Length<>22 Then
				MessageBox.Show("返回数据长度错误")
				txtTid.Text = ""
				Return
			End If
			'Challenge：6个字节，Tags Shortened TID：8个字节，Tag Response：8个字节

			txtChallenge.Text = DataConvert.ByteArrayToHexString(BLEDeviceAPI.Utils.CopyArray(result, 0, 6))
			txtTid.Text = DataConvert.ByteArrayToHexString(BLEDeviceAPI.Utils.CopyArray(result, 6, 8))
			txtResponse.Text = DataConvert.ByteArrayToHexString(BLEDeviceAPI.Utils.CopyArray(result, 14, 8))
			' MessageBox.Show("success");
		End Sub

		Private Sub txtFilter_EPC_TextChanged(ByVal sender As Object, ByVal e As EventArgs) Handles txtFilter_EPC.TextChanged
			If Not Me.txtFilter_EPC.IsHandleCreated Then Return

			FormatHex(txtFilter_EPC)
			Dim data As String = txtFilter_EPC.Text.Replace(" ", "")
			If data.Length > 0 Then
				txtFilter_EPCLen.Text = ((data.Length \ 2) + (If((data.Length Mod 2) = 0, 0, 1))).ToString() ' txtRead_Length.Text = ((data.Length / 4) + ((data.Length % 4) == 0 ? 0 : 1)).ToString();
			Else
				txtFilter_EPCLen.Text = "0"
			End If
		End Sub
		Private Sub FormatHex(ByVal txt As TextBox)
			If isDelete Then
				Return
			End If
			Dim data As String = txt.Text.Trim().Replace(" ", "")
			If data <> String.Empty Then
				Dim SelectIndex As Integer = txt.SelectionStart - 1
				Dim charData() As Char = data.ToCharArray(0, data.Length)
				Dim newChar(charData.Length - 1) As Char
				Dim index As Integer = 0
				For k As Integer = 0 To charData.Length - 1
					If StringUtils.IsHexNumber2(charData(k)) Then
						newChar(index) = charData(k)
						index += 1
					End If
				Next k
				Dim newData As New String(newChar, 0, index)
				Dim sb As New StringBuilder()
				Dim count As Integer = (newData.Length \ 2) + (newData.Length Mod 2)

				For k As Integer = 0 To count - 1
					If (k * 2 + 2) <= newData.Length Then
						sb.Append(newData.Substring(k * 2, 2))
					Else
						sb.Append(newData.Substring(k * 2, 1))
					End If
					sb.Append(" ")
				Next k
				txt.Text = sb.ToString()

				If SelectIndex >= 0 Then
					txt.SelectionStart = SelectIndex + 2
				End If
				'txt.Select(txt.Text.Length - 1, 0);

			End If
		End Sub
		Private isDelete As Boolean = False

		Private Sub M775AuthenticateForm_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs) Handles Me.KeyDown
			If e.KeyCode = Keys.Back Then
				isDelete = True
			Else
				isDelete = False
			End If
		End Sub
		Private Function HexFormat(ByVal hex As String) As String
			Try
				Dim sb As New StringBuilder()
				For k As Integer = 0 To (hex.Length \ 2) - 1
					sb.Append(hex.Substring(k * 2, 2))
					If k < hex.Length \ 2 - 1 Then
						sb.Append(" ")
					End If
				Next k
				Return sb.ToString().ToUpper()
			Catch ex As System.Exception
				Return String.Empty
			End Try

		End Function

		Private Function DetectionFiltration() As Boolean

			If txtLen.Text.Trim().Length <> 0 Then
				If Integer.Parse(txtLen.Text) > 0 Then
					Dim filerData As String = txtFilter_EPC.Text.Replace(" ", "")
					If Not StringUtils.IsHexNumber(filerData) Then
						MessageBox.Show(If(Common.isEnglish, "Please input the hex filter data!", "请输入十六进制过滤数据!"))
						Return False
					End If

					Dim filerBuff() As Byte = DataConvert.HexStringToByteArray(filerData)
					Dim filerPtr As Integer = Integer.Parse(txtPtr.Text)
					Dim filerLen As Integer = Integer.Parse(txtLen.Text)

					If (filerLen \ 8 + (If(filerLen Mod 8 = 0, 0, 1))) * 2 > filerData.Length Then
						MessageBox.Show(If(Common.isEnglish, "filter data length error!", "过滤数据和长度不匹配!")) 'to do
						Return False
					End If
				End If

			End If
			Return True
		End Function
		Private Sub rbEPC_Click(ByVal sender As Object, ByVal e As EventArgs) Handles rbEPC.Click
			If Not Me.rbEPC.IsHandleCreated Then Return

			txtPtr.Text = "32"
		End Sub

		Private Sub rbTID_Click(ByVal sender As Object, ByVal e As EventArgs) Handles rbTID.Click
			If Not Me.rbTID.IsHandleCreated Then Return

			txtPtr.Text = "0"
		End Sub

		Private Sub rbUser_Click(ByVal sender As Object, ByVal e As EventArgs) Handles rbUser.Click
			If Not Me.rbUser.IsHandleCreated Then Return

			txtPtr.Text = "0"
		End Sub

		Private Sub txtAuthenticateData_TextChanged(ByVal sender As Object, ByVal e As EventArgs) Handles txtAuthenticateData.TextChanged
			If Not Me.txtAuthenticateData.IsHandleCreated Then Return

			FormatHex(txtAuthenticateData)
			Dim data As String = txtAuthenticateData.Text.Replace(" ", "")
			If data.Length > 0 Then
				txtDataLen.Text = ((data.Length \ 2) + (If((data.Length Mod 2) = 0, 0, 1))).ToString() ' txtRead_Length.Text = ((data.Length / 4) + ((data.Length % 4) == 0 ? 0 : 1)).ToString();
			Else
				txtDataLen.Text = "0"
			End If
		End Sub
	End Class
End Namespace
